Notebook de revision de daos
El objetivo de este notebook es llevar a cabo  una primera aproximacion a los datos para comprender un poco mejor como podremos procesarlos y sacar valor de la informacion que obtendremos. Para ello se desarrollaran tareas como:

1. Identificar valores faltantes o nulos: Examinar el dataset en busca de valores faltantes o nulos en las distintas columnas. Esto es crucial, ya que los datos faltantes pueden afectar la calidad del análisis y los resultados finales. 

2. Verificar la coherencia de los datos: Es fundamental comprobar si los datos son coherentes y siguen reglas lógicas y restricciones predefinidas. Por ejemplo, si se tiene una columna de edades, se puede verificar si no hay valores negativos o si no hay registros con edades extremadamente altas o bajas.

3. Detectar valores atípicos o anomalías: Los datos pueden contener valores atípicos o anomalías que están muy alejados del rango esperado o que no siguen el patrón general de los datos. 

4. Evaluar la consistencia de los datos: Es importante asegurarse de que los datos sean consistentes en todo el conjunto de datos. Esto implica verificar que los datos cumplan con las reglas y definiciones establecidas para cada atributo.

5. Validar la calidad de las etiquetas o categorías si el conjunto de datos contiene etiquetas o categorías asignadas a cada instancia. Esto implica verificar si todas las categorías están correctamente definidas, si hay categorías redundantes o si algunas instancias tienen etiquetas incorrectas o ambiguas.

Importanto librerias necesarias

In [1]:
#! pip install pandas
import pandas as pd
import psycopg2
from pandas_profiling import ProfileReport




C:\Users\USER\AppData\Local\Temp\ipykernel_1140\3819692098.py:4: DeprecationWarning: `import pandas_profiling` is going to be deprecated by April 1st. Please use `import ydata_profiling` instead.
  from pandas_profiling import ProfileReport


Conexion a la base local de postgres

In [2]:
#parametros de conexion
host = 'localhost'
port = '5432'
database = 'DataScientistR5'
user = 'postgres'
password = 'nico1998'

#ejecutar conexion
conn = psycopg2.connect(
    host=host,
    port=port,
    database=database,
    user=user,
    password=password
)

nombre_tabla = "fraudes"

# Consulta 
query = f"SELECT DISTINCT \
    Monthh, \
    WeekOfMonth, \
    DayOfWeek, \
    Make, \
    AccidentArea, \
    NULLIF(DayOfWeekClaimed, '0') AS DayOfWeekClaimed, \
    NULLIF(MonthClaimed, '0') AS MonthClaimed, \
    WeekOfMonthClaimed, \
    Sex, \
    MaritalStatus, \
    NULLIF(Age, 0) AS Age, \
    Fault, \
    PolicyType, \
    VehicleCategory, \
    VehiclePrice, \
    FraudFound_P, \
    PolicyNumber, \
    RepNumber, \
    Deductible, \
    DriverRating, \
    CASE \
        WHEN Days_Policy_Accident = 'none' THEN '0' \
        ELSE Days_Policy_Accident \
    END AS Days_Policy_Accident, \
    CASE \
        WHEN Days_Policy_Claim = 'none' THEN '0' \
        ELSE Days_Policy_Claim \
    END AS Days_Policy_Claim, \
    CASE \
        WHEN PastNumberOfClaims = 'none' THEN '0' \
        ELSE PastNumberOfClaims \
    END AS PastNumberOfClaims, \
    AgeOfVehicle, \
    AgeOfPolicyHolder, \
    CASE \
        WHEN PoliceReportFiled = 'Yes' THEN 1 \
        WHEN PoliceReportFiled = 'No' THEN 0 \
        ELSE NULL \
    END AS PoliceReportFiled, \
    CASE \
        WHEN WitnessPresent = 'Yes' THEN 1 \
        WHEN WitnessPresent = 'No' THEN 0 \
        ELSE NULL \
    END AS WitnessPresent, \
    AgentType, \
    CASE \
        WHEN NumberOfSuppliments = 'none' THEN '0' \
        ELSE NumberOfSuppliments\
    END AS NumberOfSuppliments, \
    AddressChange_Claim, \
    NumberOfCars, \
    Yearr, \
    BasePolicy \
FROM \
    {nombre_tabla}"


# Leer la tabla en un DataFrame de pandas
df = pd.read_sql(query, conn)

# fin de la conexion
conn.close()










C:\Users\USER\AppData\Local\Temp\ipykernel_1140\3102342792.py:79: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql(query, conn)


In [3]:
df.head(20)

monthh  weekofmonth  dayofweek       make accidentarea dayofweekclaimed  \
0     Jul            2     Friday     Toyota        Urban         Thursday   
1     May            4    Tuesday    Pontiac        Urban        Wednesday   
2     Jan            2   Thursday     Toyota        Urban         Thursday   
3     Oct            3   Thursday    Pontiac        Urban           Monday   
4     Aug            2     Monday    Pontiac        Urban          Tuesday   
5     Jun            2   Saturday      Honda        Urban           Monday   
6     Nov            1     Friday     Nisson        Urban        Wednesday   
7     Apr            3     Monday     Toyota        Urban          Tuesday   
8     Mar            4     Friday      Mazda        Urban           Monday   
9     Nov            2  Wednesday      Mazda        Rural         Thursday   
10    Sep            5  Wednesday    Pontiac        Rural           Monday   
11    Mar            1   Saturday      Honda        Urban          Tuesday   
12    Jul            3     Monday  Chevrolet        Urban          Tuesday   
13    May            3    Tuesday      Mazda        Urban        Wednesday   
14    Jan            3  Wednesday     Toyota        Urban           Monday   
15    Dec            3     Monday      Honda        Urban         Thursday   
16    Mar            3     Sunday      Mazda        Urban           Friday   
17    Jun            4    Tuesday      Honda        Urban           Friday   
18    Feb            2  Wednesday      Dodge        Urban          Tuesday   
19    Mar            1   Thursday    Pontiac        Urban         Thursday   

   monthclaimed  weekofmonthclaimed     sex maritalstatus  ...  ageofvehicle  \
0           Jul                   3    Male       Married  ...       7 years   
1           May                   5  Female        Single  ...       7 years   
2           Jan                   2    Male       Married  ...   more than 7   
3           Oct                   4    Male       Married  ...       7 years   
4           Aug                   2    Male       Married  ...       6 years   
5           Jun                   2  Female       Married  ...       7 years   
6           Nov                   2    Male       Married  ...   more than 7   
7           Apr                   4    Male       Married  ...       6 years   
8           Mar                   5    Male       Married  ...   more than 7   
9           Nov                   2    Male       Married  ...       7 years   
10          Nov                   2    Male      Divorced  ...   more than 7   
11          Mar                   2    Male        Single  ...           new   
12          Jul                   3    Male        Single  ...       7 years   
13          May                   3    Male       Married  ...       7 years   
14          Jan                   5    Male        Single  ...       7 years   
15          Dec                   5    Male       Married  ...       6 years   
16          Mar                   4  Female        Single  ...       4 years   
17          Jul                   2    Male       Married  ...       7 years   
18          Feb                   3    Male       Married  ...       7 years   
19          Mar                   3  Female       Married  ...       7 years   

   ageofpolicyholder policereportfiled witnesspresent agenttype  \
0           36 to 40                 0              0  External   
1           36 to 40                 0              0  External   
2           51 to 65                 0              0  External   
3           31 to 35                 0              0  External   
4           31 to 35                 0              0  External   
5           36 to 40                 0              0  External   
6           41 to 50                 0              0  External   
7           31 to 35                 0              0  External   
8           41 to 50                 0              0  External   
9           3

Cantidad de datos vacios por columna

In [4]:
cantidad_nulos = df.isnull().sum()

# Mostrar la cantidad de valores nulos por columna
print(cantidad_nulos)

monthh                    0
weekofmonth               0
dayofweek                 0
make                      0
accidentarea              0
dayofweekclaimed          1
monthclaimed              1
weekofmonthclaimed        0
sex                       0
maritalstatus             0
age                     320
fault                     0
policytype                0
vehiclecategory           0
vehicleprice              0
fraudfound_p              0
policynumber              0
repnumber                 0
deductible                0
driverrating              0
days_policy_accident      0
days_policy_claim         0
pastnumberofclaims        0
ageofvehicle              0
ageofpolicyholder         0
policereportfiled         0
witnesspresent            0
agenttype                 0
numberofsuppliments       0
addresschange_claim       0
numberofcars              0
yearr                     0
basepolicy                0
dtype: int64


Informacion de las columnas

In [5]:
df.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15420 entries, 0 to 15419
Data columns (total 33 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   monthh                15420 non-null  object 
 1   weekofmonth           15420 non-null  int64  
 2   dayofweek             15420 non-null  object 
 3   make                  15420 non-null  object 
 4   accidentarea          15420 non-null  object 
 5   dayofweekclaimed      15419 non-null  object 
 6   monthclaimed          15419 non-null  object 
 7   weekofmonthclaimed    15420 non-null  int64  
 8   sex                   15420 non-null  object 
 9   maritalstatus         15420 non-null  object 
 10  age                   15100 non-null  float64
 11  fault                 15420 non-null  object 
 12  policytype            15420 non-null  object 
 13  vehiclecategory       15420 non-null  object 
 14  vehicleprice          15420 non-null  object 
 15  fraudfound_p       

Revisando  si hay duplicados observamos que no los hay ya que en la consulta nos aseguramos que fuera así

In [6]:

df.duplicated().value_counts() 

False    15420
Name: count, dtype: int64

Revisando y contando valores distintos en las columnas para identificar si contienen valores anormales como 0.

In [7]:
for columna in df:
    if columna!='policynumber':
        print(columna)
        print(df[columna].unique(),"\n")

monthh
['Jul' 'May' 'Jan' 'Oct' 'Aug' 'Jun' 'Nov' 'Apr' 'Mar' 'Sep' 'Dec' 'Feb'] 

weekofmonth
[2 4 3 1 5] 

dayofweek
['Friday' 'Tuesday' 'Thursday' 'Monday' 'Saturday' 'Wednesday' 'Sunday'] 

make
['Toyota' 'Pontiac' 'Honda' 'Nisson' 'Mazda' 'Chevrolet' 'Dodge' 'VW'
 'Ford' 'Accura' 'BMW' 'Saturn' 'Saab' 'Ferrari' 'Mercury' 'Mecedes'
 'Lexus' 'Porche' 'Jaguar'] 

accidentarea
['Urban' 'Rural'] 

dayofweekclaimed
['Thursday' 'Wednesday' 'Monday' 'Tuesday' 'Friday' 'Saturday' 'Sunday'
 None] 

monthclaimed
['Jul' 'May' 'Jan' 'Oct' 'Aug' 'Jun' 'Nov' 'Apr' 'Mar' 'Dec' 'Feb' 'Sep'
 None] 

weekofmonthclaimed
[3 5 2 4 1] 

sex
['Male' 'Female'] 

maritalstatus
['Married' 'Single' 'Divorced' 'Widow'] 

age
[36. 39. 57. 30. 26. 42. 47. 51. 44. 48. nan 38. 29. 32. 40. 25. 53. 34.
 71. 41. 37. 50. 43. 52. 74. 35. 21. 49. 33. 58. 45. 67. 28. 31. 59. 22.
 65. 61. 60. 54. 24. 63. 27. 73. 55. 66. 56. 18. 62. 46. 80. 70. 23. 20.
 68. 64. 76. 72. 69. 75. 19. 78. 77. 79. 17. 16.] 

fault
['Policy Hol

In [8]:
# Conteo de valores de la columna fraudfound_p
conteo_valores = df['fraudfound_p'].value_counts().sort_index()
print(conteo_valores.sum)

<bound method NDFrame._add_numeric_operations.<locals>.sum of fraudfound_p
0    14497
1      923
Name: count, dtype: int64>


In [9]:
# Conteo de valores de la columna day of week claimed, observamos que en esta columna hay un valor que contiene un 0 
#cuando esta columna solo deberia contener nombres de días , debemos corregir este error.
conteo_valores = df['dayofweekclaimed'].value_counts(dropna=False)

# Mostrar el conteo de valores
print(conteo_valores)

dayofweekclaimed
Monday       3757
Tuesday      3375
Wednesday    2951
Thursday     2660
Friday       2497
Saturday      127
Sunday         52
None            1
Name: count, dtype: int64


Revisando mas a profundidad el registro en 0 de 'dayofweekclaimed'

In [10]:
df[(df['dayofweekclaimed'].isnull())]


monthh  weekofmonth dayofweek   make accidentarea dayofweekclaimed  \
8252    Jul            2    Monday  Honda        Rural             None   

     monthclaimed  weekofmonthclaimed   sex maritalstatus  ...  ageofvehicle  \
8252         None                   1  Male        Single  ...           new   

     ageofpolicyholder policereportfiled witnesspresent agenttype  \
8252          16 to 17                 0              0  External   

      numberofsuppliments  addresschange_claim  numberofcars  yearr  \
8252                    0            no change     1 vehicle   1994   

      basepolicy  
8252  All Perils  

[1 rows x 33 columns]

In [11]:
# Conteo de valores de la columna MONTHCLAIMED observamos que en esta columna hay un valor que contiene un 0 
#cuando esta columna solo deberia contener nombres de meses , debemos corregir este error.
conteo_valores = df['monthclaimed'].value_counts(dropna=False)

# Mostrar el conteo de valores
print(conteo_valores)

monthclaimed
Jan     1446
May     1411
Mar     1348
Oct     1339
Jun     1293
Feb     1287
Nov     1285
Apr     1271
Sep     1242
Jul     1225
Dec     1146
Aug     1126
None       1
Name: count, dtype: int64


In [12]:
df[(df['monthclaimed'].isnull())]


monthh  weekofmonth dayofweek   make accidentarea dayofweekclaimed  \
8252    Jul            2    Monday  Honda        Rural             None   

     monthclaimed  weekofmonthclaimed   sex maritalstatus  ...  ageofvehicle  \
8252         None                   1  Male        Single  ...           new   

     ageofpolicyholder policereportfiled witnesspresent agenttype  \
8252          16 to 17                 0              0  External   

      numberofsuppliments  addresschange_claim  numberofcars  yearr  \
8252                    0            no change     1 vehicle   1994   

      basepolicy  
8252  All Perils  

[1 rows x 33 columns]

In [13]:
# Conteo de valores de la columna AGE observamos que en esta columna hay 320 valores que contiene 0 
#cuando esta columna solo deberia contener edades validas ya que representa la edad
#de quien se vio incolucrado en el accidente y presento la denuncua , debemos corregir este error ya sea
#eliminando estos regitros o con alguna otra estrategia.
conteo_valores = df['age'].value_counts().sort_index()

# Mostrar el conteo de valores
print(conteo_valores)

age
16.0     9
17.0     6
18.0    48
19.0    32
20.0    28
        ..
76.0    42
77.0    29
78.0    35
79.0    20
80.0    32
Name: count, Length: 65, dtype: int64


Revisando mas a profundidad la columna 'AGE'

In [14]:
df[(df['age'].isnull())]['fraudfound_p'].value_counts()


fraudfound_p
0    289
1     31
Name: count, dtype: int64

Conclusiones preliminares de las variables 'AGE' , 'MONTHCLAIMED' Y 'DAYOFWEEKCLAIMED':

1. Debido a que la variable MONTHCLAIMED Y DAY OF WEEK CLAIMED tan solo tienen un registro con valor nulo y dicho registro es el mismo, podemos borrar este registro sin afectar los datos.

2. La columna 'AGE' tiene 320 registros con edade nulas, para solucionar este caso hay varios procedimientos que se pueden tomar:
    * Borrar estos 320 registros ya que tan solo representan 320/15420*100= 2.007% de los datos , sin embargo es necesario revisar
    como se encuentra la variable objetivo distribuida en estos registros. Al hacer esto observamos que en los registros con 'AGE'= 0 
    se encuentran 289 casos de reclamaciones no fraudulentas y 31 casos de reclamaciones fraudulentas. Esto quiere decir que en los registros
    con 'AGE'= 0 se encuentra el 289/14497 x 100= 2% de transacciones no fraudulentas y el  31/923 x 100 =3.52%  de transacciones fraudulentas

    * Otra solucion seria imputar la variable AGE con la variable AGEOFPOLICYHOLDER. sin embargo la variable AGE es numerica y AGEPOLICYHOLDER 
    es categorica. Esta solucion requeriria  crear una variable adicional llamada MEAN_AGEOFPOLICYHOLDER almacenando el
    promedio entre el rango superior e inferior de la variable AGEOFPOLICYHOLDER e imputar  la columna AGE usando esta nueva columna.

    * Otra opcion seria usar un modelo de machine learning para imputar estas edades usando las demas caracteristicas para crear clusters de datos que
    permitan asignar la edad, sin embargo esta estrategia requeriria un tratamiento mas complejo.



Limpiando las variables  'MONTHCLAIMED' Y 'DAYOFWEEKCLAIMED':

In [15]:
#Quitando del dataset los registros que tienen MONTHCLAIMED='0'
df = df[~(df['monthclaimed'].isnull())]
df = df[~(df['dayofweekclaimed'].isnull())]



Revisando dataframe despues de quitar los registros con monthclaimed y dayofweekclaimed


In [16]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 15419 entries, 0 to 15419
Data columns (total 33 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   monthh                15419 non-null  object 
 1   weekofmonth           15419 non-null  int64  
 2   dayofweek             15419 non-null  object 
 3   make                  15419 non-null  object 
 4   accidentarea          15419 non-null  object 
 5   dayofweekclaimed      15419 non-null  object 
 6   monthclaimed          15419 non-null  object 
 7   weekofmonthclaimed    15419 non-null  int64  
 8   sex                   15419 non-null  object 
 9   maritalstatus         15419 non-null  object 
 10  age                   15100 non-null  float64
 11  fault                 15419 non-null  object 
 12  policytype            15419 non-null  object 
 13  vehiclecategory       15419 non-null  object 
 14  vehicleprice          15419 non-null  object 
 15  fraudfound_p          15

Revisando balance de los datos en la variable objetivo (fraudfound_p) observamos que tenemos un desbalanceo de clase fuerte
ya que hay 14208 registros que se identificaron como no fraudulentos y tan solo 892 identificados como fraudulentos, es importante tener esto en
cuenta a la hora de construir los modelos de machine learning.


In [17]:
# Conteo de valores de la columna fraudfound_p
conteo_valores = df['fraudfound_p'].value_counts().sort_index()

# Mostrar el conteo de valores
print(conteo_valores.sum)

<bound method NDFrame._add_numeric_operations.<locals>.sum of fraudfound_p
0    14496
1      923
Name: count, dtype: int64>
